### PR TITLE
PPA Collect key submission metadata (EXPOSUREAPP-4821)

### PR DESCRIPTION
**Pull request by chiljamgossow**
Tech spec: https://github.com/corona-warn-app/cwa-app-tech-spec/blob/40e00dfd0e4544bdf05a2ae83fa7123af86b9438/docs/spec/data-donation-client-ppa.md#key-submission-metadata

How to test:
- go through the submission flow and check whether the data donation values match your behavior

Test cases:
- with or without prior high risk warning
- with or without prior consent
- with or without withdrawal of consent
- ending the submission flow on different screens
- trigger background submission by not finishing the symptom flow
- use QR code or TAN

Missing: Unit tests

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionCollector.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    private val riskLevelStorage: RiskLevelStorage,
+    private val riskLevelSettings: RiskLevelSettings
+) {
+
+    fun reset() {
+        analyticsKeySubmissionStorage.clear()
+    }
+
+    fun reportPositiveTestResultReceived() {
+        if (isEnabled) analyticsKeySubmissionStorage.testResultReceivedAt.update { timeStamper.nowUTC.millis }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionCollector.kt:27]: Not a blocker, but personally, I would prefer a simple check using the same pattern everywhere, no double negations, or different checks of `isEnabled` vs `isDisabled`, i.e. early aborts with the same pattern.

Everywhere just `if(!isEnabled) return`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionDonor.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    private fun shouldSubmitData(timeSinceTestResultToSubmit: Duration): Boolean {
+        return positiveTestResultReceived && (
+            keysSubmitted || enoughTimeHasPassedSinceResult(timeSinceTestResultToSubmit)
+            )
+    }
+
+    private val positiveTestResultReceived
+        get() = repository.testResultReceivedAt > 0
+
+    private val keysSubmitted
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionDonor.kt:66]: Please add return type, it's not possible to see what it is when viewing without IDE.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import org.joda.time.Instant
+import javax.inject.Inject
+import kotlin.math.max
+
+class AnalyticsKeySubmissionRepository @Inject constructor(
+    private val storage: AnalyticsKeySubmissionStorage,
+    private val timeStamper: TimeStamper,
+    private val riskLevelSettings: RiskLevelSettings
+) {
+    val testResultReceivedAt
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt:15]: Please add return type.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        get() = storage.testResultReceivedAt.value
+
+    private val testRegisteredAt: Long
+        get() = storage.testRegisteredAt.value
+
+    val submitted
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt:21]: Please add return type.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        get() = storage.submitted.value
+
+    val submittedInBackground
+        get() = submitted && storage.submittedInBackground.value
+
+    val submittedAfterCancel
+        get() = submitted && storage.submittedAfterCancel.value
+
+    val submittedAfterSymptomFlow
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt:30]: Please add return type.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        get() = storage.submittedAfterSymptomFlow.value
+
+    val submittedWithTeleTAN
+        get() = submitted && storage.registeredWithTeleTAN.value
+
+    val lastSubmissionFlowScreen
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionRepository.kt:36]: Please add return type.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt [hunk 7]
```diff
@@ -121,17 +123,20 @@ class SubmissionRepository @Inject constructor(
     suspend fun asyncRegisterDeviceViaTAN(tan: String) {
         val registrationData = submissionService.asyncRegisterDeviceViaTAN(tan)
         LocalData.registrationToken(registrationData.registrationToken)
-        updateTestResult(registrationData.testResult)
         LocalData.devicePairingSuccessfulTimestamp(timeStamper.nowUTC.millis)
+        updateTestResult(registrationData.testResult)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt:127]: Why was the call order changed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt [hunk 8]
```diff
@@ -121,17 +123,20 @@ class SubmissionRepository @Inject constructor(
...
+        analyticsKeySubmissionCollector.reportTestRegistered()
+        analyticsKeySubmissionCollector.reportRegisteredWithTeleTAN()
     }
 
     suspend fun asyncRegisterDeviceViaGUID(guid: String): TestResult {
         val registrationData = submissionService.asyncRegisterDeviceViaGUID(guid)
         LocalData.registrationToken(registrationData.registrationToken)
-        updateTestResult(registrationData.testResult)
         LocalData.devicePairingSuccessfulTimestamp(timeStamper.nowUTC.millis)
+        updateTestResult(registrationData.testResult)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt:137]: Why was the call order changed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/task/SubmissionTask.kt [hunk 9]
```diff
@@ -48,9 +51,13 @@ class SubmissionTask @Inject constructor(
...
 
-            if (arguments.checkUserActivity && hasRecentUserActivity()) {
-                Timber.tag(TAG).w("User has recently been active in submission, skipping submission.")
-                return Result(state = Result.State.SKIPPED)
+            if (arguments.checkUserActivity) {
+                if (hasRecentUserActivity()) {
+                    Timber.tag(TAG).w("User has recently been active in submission, skipping submission.")
+                    return Result(state = Result.State.SKIPPED)
+                } else {
+                    analyticsKeySubmissionCollector.reportSubmittedInBackground()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/task/SubmissionTask.kt:59]: Submission may still abort for other reasons, should we call `reportSubmittedInBackground` later when we are sure that it was submitted?

File: Corona-Warn-App/src/main/res/drawable/contact_diary_duration_background_selected.xml [hunk 10]
```diff
@@ -4,4 +4,4 @@
     <corners android:radius="@dimen/radius_card" />
     <solid android:color="@color/colorSurface1" />
     <stroke android:width="2dip" android:color="@color/colorAccent" />
-</shape>
\ No newline at end of file
+</shape>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/drawable/contact_diary_duration_background_selected.xml:7]: Unrelated?

File: Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml [hunk 11]
```diff
@@ -135,4 +135,4 @@
         android:name="de.rki.coronawarnapp.ui.information.InformationTermsFragment"
         android:label="@layout/fragment_information_terms"
         tools:layout="@layout/fragment_information_terms" />
-</navigation>
\ No newline at end of file
+</navigation>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml:138]: Unrelated?

**Review by d4rken — COMMENTED** [review]
Code looks good, few comments that are not blockers.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt [hunk 12]
```diff
@@ -32,6 +33,7 @@ class SubmissionConsentViewModel @AssistedInject constructor(
         launch {
             try {
                 val preAuthorized = tekHistoryProvider.preAuthorizeExposureKeyHistory()
+                if (preAuthorized) analyticsKeySubmissionCollector.reportAdvancedConsentGiven()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt:36]: AFAIK 

> advancedConsentGiven set to true if the user agreed to share keys when scanning the QR code and the user has not withdrawn the consent; false otherwise.

is about the legal consent, i.e.

Screenshot from 2021-02-26 16-55-02

not about the technical consent  to access the TEKs.

ping @mlenkeit 
cc @mtwalli

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt [hunk 13]
```diff
@@ -59,6 +61,7 @@ class SubmissionConsentViewModel @AssistedInject constructor(
 
     fun giveGoogleConsentResult(accepted: Boolean) {
         Timber.i("User allowed Google consent:$accepted")
+        if (accepted) analyticsKeySubmissionCollector.reportAdvancedConsentGiven()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt:64]: Legal vs Technical consent?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/resultavailable/SubmissionTestResultAvailableViewModel.kt [hunk 14]
```diff
@@ -51,6 +53,7 @@ class SubmissionTestResultAvailableViewModel @AssistedInject constructor(
 
             override fun onTEKPermissionDeclined() {
                 Timber.d("onTEKPermissionDeclined")
+                analyticsKeySubmissionCollector.reportConsentWithdrawn()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/resultavailable/SubmissionTestResultAvailableViewModel.kt:56]: Legal vs Technical consent?

**Review by d4rken — CHANGES_REQUESTED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt (hunk 12 above)
Comment (mlenkeit) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt:36]: @d4rken is right. Sorry for not being particularly clear in the spec. When it was written, the "technical consent" didn't even exist I think 😄

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionCollector.kt (hunk 1 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionCollector.kt]: I prefer if (disabled). Negation is so hard to read.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt (hunk 7 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt:126]: Not necessary anymore. The implementation is independent from the order.

File: Corona-Warn-App/src/main/res/drawable/contact_diary_duration_background_selected.xml (hunk 10 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/res/drawable/contact_diary_duration_background_selected.xml:7]: I tried to revert several times, but git just says no changes detected. I do not know why this appears here.

File: Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml (hunk 11 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml:138]: same here

File: Corona-Warn-App/src/main/res/drawable/contact_diary_duration_background_selected.xml (hunk 10 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/res/drawable/contact_diary_duration_background_selected.xml:7]: Accidentally selected "ignore whitespace changes"?

**Review by d4rken — APPROVED** [review]
lgtm

**Comment by LukasLechnerDev** [comment]
I made some tests on my Samsung S20+: 

**Test 1:** 
1. Without prior high-risk warning
2. Full Submission Flow, Get back to Home Screen with "Vielen Dank!"
3. Not triggering Background Submission => Entering Symptoms
4. QR-Code

Analytics Data: 
```
key_submission_metadata_set {
  hours_since_high_risk_warning_at_test_registration: -1
  last_submission_flow_screen: SUBMISSION_FLOW_SCREEN_SYMPTOM_ONSET
  last_submission_flow_screen_value: 5
  submitted: true
  submitted_after_symptom_flow: true
}
```

Shared Prefs: 
```
<map>
    <long name="analytics_key_submission_testRegisteredAt" value="1614590994921" />
    <int name="analytics_key_submission_lastSubmissionFlowScreen" value="5" />
    <boolean name="analytics_key_submission_submitted" value="true" />
    <boolean name="analytics_key_submission_submittedAfterSymptomFlow" value="true" />
    <long name="analytics_key_submission_testResultReceivedAt" value="1614590994917" />
    <int name="analytics_key_submission_riskLevelAtTestRegistration" value="2" />
</map>
```

**Test 2:** 
1. With a prior high-risk warning
2. Full Submission Flow without entering Symptoms, Get back to Home Screen with "Vielen Dank!"
3. Triggering Background Submission => Not Entering Symptoms
4. Using Tele-TAN

Analytics Data: 
```
key_submission_metadata_set {
  last_submission_flow_screen: SUBMISSION_FLOW_SCREEN_WARN_OTHERS
  last_submission_flow_screen_value: 3
  submitted: true
  submitted_with_tele_t_a_n: true
}
```

Shared Preferences:
```
<map>
    <long name="analytics_key_submission_testRegisteredAt" value="1614595414888" />
    <int name="analytics_key_submission_hoursSinceHighRiskWarningAtTestRegistration" value="0" />
    <boolean name="analytics_key_submission_submitted" value="true" />
    <int name="analytics_key_submission_lastSubmissionFlowScreen" value="3" />
    <boolean name="analytics_key_submission_registeredWithTeleTAN" value="true" />
    <long name="analytics_key_submission_testResultReceivedAt" value="1614595414885" />
    <int name="analytics_key_submission_riskLevelAtTestRegistration" value="3" />
</map>
```

**Test 3**
I tested the `lastSubmissionFlowScreen` property by entering the different screens and checking if the respective value go stored in the shared preferences and this worked fine for all screens. 


So I haven't found any issues in my tests. 
Good work 👍

**Review by LukasLechnerDev — APPROVED** [review]
